### PR TITLE
GPU: Track buffer migrations and flush source on incomplete copy

### PR DESCRIPTION
**Pull request by riperiperi**
Makes sure that the modified range list is always from the latest iteration of the buffer, and flushes earlier iterations of a buffer if the data has not been migrated yet.

Before, flushes on migrated buffers could flush uninitialized data, as it would assume that a sync number for a previous write had been reached, but the copy to migrate the buffer data had not yet been reached. The flush's change to the modified range could also be lost when switching, as the change would be applied to the dead buffer.

Buffers now have a reference count to keep their data alive. When a migration is performed (with modified ranges), then it takes a reference to the source buffer and gets added to a list on `_context`. When any sync is created, we check the current reached sync number and pending migrations to see if any can be removed. If they can, 

Upsides:
- Avoids this cause of buffer data loss.
Downsides:
- Buffers with modifications are kept alive after migrations until the next sync number is reached. This could cause increased memory spikes when resizing often.
  - Buffers without modifications don't have any issues. Maybe this could be improved by force flushing buffer data where the sync number has been reached, but I'm not sure about the performance implications (on opengl this isn't as free, when vulkan is choosing buffer memory types we don't want our cleanup to influence that either).
- Can't reuse BufferModifiedRangeList when migrating, as we need to track old states for flush.

This fixes most cases of vertex explosions in Pokemon Scarlet/Violet. On master without vsync, going through Mesagoza to the school, then to the east loading zone caused the main character to explode and never recover very consistently. It took hours of running around with vsync off to get an explosion with these changes, and it was on NPCs, and they recovered when they were reloaded. There may be some other issue with buffer flush that might be causing these.

### CODE REVIEW

File: Ryujinx.Graphics.OpenGL/Sync.cs [hunk 1]
```diff
@@ -40,6 +40,37 @@ public void Create(ulong id)
...
+                    lock (handle)
+                    {
+                        if (handle.Handle == IntPtr.Zero)
+                        {
+                            continue;
+                        }
+
+                        if (handle.ID > lastHandle)
+                        {
+                            WaitSyncStatus syncResult = GL.ClientWaitSync(handle.Handle, _syncFlags, 1000000000);
```
[the remark this answers is not in the document]
Comment (riperiperi) [reply Ryujinx.Graphics.OpenGL/Sync.cs:60]: Oh, that's not intentional at all.

File: Ryujinx.Graphics.Gpu/Memory/BufferModifiedRangeList.cs [hunk 2]
```diff
@@ -313,44 +431,20 @@ public void InheritRanges(BufferModifiedRangeList ranges, Action<ulong, ulong> r
...
-
-            ulong currentSync = _context.SyncNumber;
-            for (int i = 0; i < count; i++)
-            {
-                BufferModifiedRange range = ranges[i];
-                if (range.SyncNumber != currentSync)
-                {
-                    rangeAction(range.Address, range.Size);
-                }
+                bool removed = _sources.Remove(migration);
```
[the remark this answers is not in the document]
Comment (riperiperi) [reply Ryujinx.Graphics.Gpu/Memory/BufferModifiedRangeList.cs:447]: This was for a debug log, which I removed. 🥴

File: Ryujinx.Graphics.Gpu/Memory/BufferModifiedRangeList.cs [hunk 3]
```diff
@@ -264,47 +402,39 @@ public void WaitForAndGetRanges(ulong address, ulong size, Action<ulong, ulong>
...
 
-            lock (_lock)
-            {
-                foreach (BufferModifiedRange range in inheritRanges)
+                lock (_lock)
                 {
-                    Add(range);
+                    (_sources ??= new List<BufferMigration>()).Add(migration);
+
+                    _sources.Add(migration);
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Memory/BufferModifiedRangeList.cs:432]: You already added it above.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks. Nice to see that problem fixed.

**Review by AcK77 — APPROVED** [review]

**Pull request merged**